pkg/autoscaler/aggregation/bucketing.go

```diff
@@ -161,31 +161,44 @@ func (t *TimedFloat64Buckets) Record(now time.Time, value float64) {
 	writeIdx := t.timeToIndex(now)
 
 	if t.lastWrite != bucketTime {
-		if t.firstWrite.IsZero() {
-			t.firstWrite = bucketTime
+		if bucketTime.Add(t.window).Before(t.lastWrite) {
+			// Ingore this value because it happened a window size ago
+			return
 		}
-		// This should not really happen, but is here for correctness.
-		if bucketTime.Sub(t.lastWrite) > t.window {
-			// This means we had no writes for the duration of `window`. So reset the firstWrite time.
+
+		// If it is the first write or it happened before the first write which we
+		// have in record, update the firstWrite.
+		if t.firstWrite.IsZero() || t.firstWrite.After(bucketTime) {
 			t.firstWrite = bucketTime
```

Comment:
Do we need to be zero-ing things here or is this guaranteed to be safe? If we move firstWrite backwards it seems like we might rotate around and reuse garbage data here - or can this not happen?

I added a test case just to play with this and it looks like there might be some bug here somewhere (or I goofed, which is also very possible!):

~~~~
func TestDescendingRecord(t *testing.T) {
	now := time.Now()
	buckets := NewTimedFloat64Buckets(5*time.Second, 1*time.Second)

	for i := 8 * time.Second; i >= 0*time.Second; i -= time.Second {
		buckets.Record(now.Add(i), 5)
	}

	if got, want := buckets.WindowAverage(now.Add(5*time.Second)), 5.; got != want {
		// we wrote a 5 every second, and we never wrote in the same second twice,
		// so the average _should_ be 5.
		t.Errorf("WindowAverage = %v, want: %v", got, want)
	}
}

// bucketing_test.go:187: WindowAverage = 6, want: 5
// printing out buckets, I see:
// ([]float64) (len=5 cap=5) {
//  (float64) 5,
//  (float64) 10,
//  (float64) 5,
//  (float64) 5,
//  (float64) 5
// }
~~~~

Reply:
Well you do write twice in the same bucket (after wraparound that is). But this looks sketchy, yes

Reply:
Same bucket, but not same second :D

Reply:
Good caught. I think we won't reuse garbage data as we skip recording if the data is too old. I think the point here is what we should do for `bucketTime+window == lastwrite` case. I did record before, which means there should be `window+1` slots, which isn't right. I changed to skip and your test passed.